### PR TITLE
Add Eq intrinsics and type class

### PR DESCRIPTION
**Pull request by esbenbjerre**
- [x] Add EQ intrinsics
- [x] Add Eq type class
- [x] Add instances for tuples
- [x] Add instances for types in std. library
- [x] Refactor functions in std. library with Eq constraint
- [x] Solve resulting type errors in library and tests

### CODE REVIEW

**Comment by esbenbjerre** [comment]
So this PR naturally introduces some type errors in existing tests, one of them is this one
```
def testEqMap01(): Bool = Map#{} == Map#{}
```
which results in
```
>> No instance of class 'Eq' for type 'k'.
...
>> No instance of class 'Eq' for type 'v'.
```
I guess the issue here is that the type of the keys and values could be anything. Is the solution below acceptable for solving this? Or do we want to do something clever?
```
def testEqMap01(): Bool =
    let m1: Map[Int32, Int32] = Map#{};
    let m2: Map[Int32, Int32] = Map#{};
    m1 == m2
```

**Comment by magnus-madsen** [comment]
pinging @mlutze:

When a type variable is "under constrained" as in Esbens example:

```
def testEqMap01(): Bool = Map#{} == Map#{}
```

What should we do?

I believe Haskell reports this as an error, so maybe we should too? Can we have a specific error?

**Comment by esbenbjerre** [comment]
I just wanted to add that it works in Haskell.
```
GHCi, version 8.10.1: https://www.haskell.org/ghc/  :? for help
Loaded GHCi configuration from /Users/esbenbjerre/.ghci
λ import Data.Map
λ let m1 = Data.Map.empty
λ let m2 = Data.Map.empty
λ m1 == m2
True
λ 
```
I was also quite surprised to find this definition of `Eq`
```haskell
{--------------------------------------------------------------------
  Eq converts the tree to a list. In a lazy setting, this 
  actually seems one of the faster methods to compare two trees 
  and it is certainly the simplest :-)
--------------------------------------------------------------------}
instance (Eq k,Eq a) => Eq (Map k a) where
  t1 == t2  = (size t1 == size t2) && (toAscList t1 == toAscList t2)
```

**Comment by magnus-madsen** [comment]
Interesting. Thanks for testing it!

**Comment by mlutze** [comment]
Yes, this is definitely doable without error. I've been waiting for the "complex/overlapping instances" discussion to settle before addressing this; if we have no overlapping instances, I think we can do this in a fairly straightforward way, so we won't need to annotate the empty maps. If overlapping instances are allowed, then things are a bit more complicated, as we potentially have to decide which Eq instance to use among multiple matches.

**Comment by magnus-madsen** [comment]
My impressions that overlapping instances are not great; they might be very handy, but they are not principles and they are potentially dangerous.

**Comment by mlutze** [comment]
Yeah, I'm leaning in that direction, too.

**Comment by esbenbjerre** [comment]
There are 100s of tests which depends on equality checks of "under constrained" values, so I think this is essentially blocked until we have a solution for this.

**Comment by magnus-madsen** [comment]
@mlutze 

For
 
```
def testEqMap01(): Bool = Map#{} == Map#{}
```

which instance would you pick? (Unit?) and why is that not dangerous?

**Comment by mlutze** [comment]
The instance immediately used is `Eq[Map[a, b]]` of course, which would presumably eventually delegate to `Eq[a]` and `Eq[b]`. However, neither of those corresponding functions would actually be called, as the map is empty, so we could use some kind of dummy function. I think this generalizes to all cases where the constrained type variables are free variables(?)

**Comment by magnus-madsen** [comment]
> 
> 
> The instance immediately used is `Eq[Map[a, b]]` of course, which would presumably eventually delegate to `Eq[a]` and `Eq[b]`. However, neither of those corresponding functions would actually be called, as the map is empty, so we could use some kind of dummy function. I think this generalizes to all cases where the constrained type variables are free variables(?)

If a program is undertyped it does not really have a meaning. The classic example is `toString(default())`. 

@esbenbjerre How many tests are we talking? In most cases it should be sufficient with a single type annotation.

Maybe some of the tests should be deleted?

Are these tests mostly in the std lib or for the compiler or a mix?

**Comment by esbenbjerre** [comment]
> @esbenbjerre How many tests are we talking? In most cases it should be sufficient with a single type annotation.
> 
> Maybe some of the tests should be deleted?
> 
> Are these tests mostly in the std lib or for the compiler or a mix?

At least 100 of the std. library tests. I have already commented out about a dozen from `LangSuite`. Something as
```
def unzip01(): Bool = Option.unzip(None) == (None, None)
```
also fails. In these cases I would have to add a "dummy" type such as `Option[Unit]` or `Option[Int32]`.

**Comment by mlutze** [comment]
Maybe Haskell's "default" mechanism applies here. But I don't like it much as "it involves making choices about the semantics of a program in ways that are not always directly visible to programmers." (THIH 11.5.1)

**Comment by magnus-madsen** [comment]
> ```
> def unzip01(): Bool = Option.unzip(None) == (None, None)
> ```
> 
> also fails. In these cases I would have to add a "dummy" type such as `Option[Unit]` or `Option[Int32]`.

Could I ask you to do - say 10 or 20 - of those and then let me review? Try to use the "minimum" number of type ascriptions required. E.g. in the above you would probably write

```
def unzip01(): Bool = Option.unzip(None: Option[(Unit, Unit)]) == (None, None)
```

since you have to choose a type, I suggest we go with `Unit` by default.

After the 10-20 updated tests we can review.

**Comment by magnus-madsen** [comment]
(If you get blocked on this- feel free to take a look at e.g. the Hash TC #1531)

**Comment by esbenbjerre** [comment]
The only "big" code change left is in `TestMap.flix`. I'll try to get around it tomorrow.

**Comment by esbenbjerre** [comment]
Hmm. There are still some type errors. In the lattice examples, e.g. `examples/domains/Belnap.flix` there are stuff like
```
equ(e1: Belnap, e2: Belnap): Bool = e1 == e2
```
which does not behave nice. I can add a proper definition, but the issues extends into the laws defined in `main/src/library/Bounded.flix`.
What do you think we should do here @magnus-madsen?

**Comment by magnus-madsen** [comment]
> 
> 
> Hmm. There are still some type errors. In the lattice examples, e.g. `examples/domains/Belnap.flix` there are stuff like
> 
> ```
> equ(e1: Belnap, e2: Belnap): Bool = e1 == e2
> ```
> 
> which does not behave nice. I can add a proper definition, but the issues extends into the laws defined in `main/src/library/Bounded.flix`.
> What do you think we should do here @magnus-madsen?

I think you need to go an implement your own equality by pattern matching on both elements and doing the "straightforward" thing.

**Comment by esbenbjerre** [comment]
I have some trouble with
```
law strict1[a, b](f: a -> b, bot: a, cobot: b): Bool = f(bot) == cobot
law strict2[a, b, c](f: (a, b) -> c, bot1: a, bot2: b, cobot: c): Bool = f(bot1, bot2) == cobot
```
 in `main/src/library/Bounded.flix`.
I've tried to add a `PreOrder` constraint and using the `equ` method, but then i get
```
Exception in thread "main" ca.uwaterloo.flix.util.InternalCompilerException: Expected exactly 1 matching signature, found 0
	at ca.uwaterloo.flix.language.phase.Monomorph$.specializeSigSym$1(Monomorph.scala:752)
	at ca.uwaterloo.flix.language.phase.Monomorph$.visitExp$1(Monomorph.scala:166)
	at ca.uwaterloo.flix.language.phase.Monomorph$.visitExp$1(Monomorph.scala:225)
        ...
``

**Comment by magnus-madsen** [comment]
> 
> 
> I have some trouble with
> 
> ```
> law strict1[a, b](f: a -> b, bot: a, cobot: b): Bool = f(bot) == cobot
> law strict2[a, b, c](f: (a, b) -> c, bot1: a, bot2: b, cobot: c): Bool = f(bot1, bot2) == cobot
> ```
> 
> in `main/src/library/Bounded.flix`.
> I've tried to add a `PreOrder` constraint and using the `equ` method, but then i get
> 
> ```
> Exception in thread "main" ca.uwaterloo.flix.util.InternalCompilerException: Expected exactly 1 matching signature, found 0
> 	at ca.uwaterloo.flix.language.phase.Monomorph$.specializeSigSym$1(Monomorph.scala:752)
> 	at ca.uwaterloo.flix.language.phase.Monomorph$.visitExp$1(Monomorph.scala:166)
> 	at ca.uwaterloo.flix.language.phase.Monomorph$.visitExp$1(Monomorph.scala:225)
>         ...
> ``
> ```

Try to check the signatures carefully. If that does not help, I think you just comment this out for now (and then you have to comment out even more code).

**Comment by esbenbjerre** [comment]
It seems like there is some implicit use of `==` in these examples. I'm not sure where the dependency comes from.
I can add a "dummy" `Eq` instance for these, the alternative is to comment out almost everything. What do you prefer?

**Comment by magnus-madsen** [comment]
> It seems like there is some implicit use of `==` in these examples. I'm not sure where the dependency comes from.
> I can add a "dummy" `Eq` instance for these (EDIT: that calls `PreOrder.equ`), the alternative is to comment out almost everything. What do you prefer?

I am not sure what you mean? I think there could be some implicit use of `==` in the Datalog constraints.

**Comment by esbenbjerre** [comment]
> I am not sure what you mean? I think there could be some implicit use of `==` in the Datalog constraints.
Let me reformulate. Can you accept if I add Eq instances which just calls PreOrder.equ?

**Comment by magnus-madsen** [comment]
> 
> 
> > I am not sure what you mean? I think there could be some implicit use of `==` in the Datalog constraints.
> > Let me reformulate. Can you accept if I add Eq instances which just calls PreOrder.equ?

Yes that is fine for now

**Comment by esbenbjerre** [comment]
This should be ready unless I've done something stupid :)
I don't know if you would like to keep `__eq` for a while or we should kill it in this PR.

**Comment by magnus-madsen** [comment]
Let us keep __eq -- for now.

**Comment by magnus-madsen** [comment]
This will take some time to review, but I already have a quick comment: 

I would just define `neq` in terms of `not eq(x, y)`. (or does that not work yet?)

Later, `neq` should just be defined in one place as `not eq(x, y)`.

**Comment by magnus-madsen** [comment]
The above comment about neq does *NOT* apply to primitive types. There we should keep both versions.

**Comment by esbenbjerre** [comment]
> This will take some time to review, but I already have a quick comment:

Understandable!

> I would just define `neq` in terms of `not eq(x, y)`. (or does that not work yet?)

I vaguely remember having an idea about making different implementations (for some types) to improve efficiency, as it's sometimes quicker to determine whether two values are not equal. But I can definitely refactor to the simpler version.

**Comment by magnus-madsen** [comment]
> I vaguely remember having an idea about making different implementations (for some types) to improve efficiency, as it's sometimes quicker to determine whether two values are not equal. But I can definitely refactor to the simpler version.

Yes, in general that can be useful, but in many cases it makes no differences.

**Comment by esbenbjerre** [comment]
> The above comment about neq does _NOT_ apply to primitive types. There we should keep both versions.

What about tuples?

**Comment by magnus-madsen** [comment]
> 
> 
> > The above comment about neq does _NOT_ apply to primitive types. There we should keep both versions.
> 
> What about tuples?

Tuples are not primitives, so just use `not eq`.

File: main/test/ca/uwaterloo/flix/library/Core/Io/TestFile.flix [hunk 1]
```diff
@@ -34,14 +34,14 @@ def exists02(): Bool & Impure = exists(new("./doc/")) |> Result.isOk
 @test
 def exists03(): Bool & Impure = exists(new("./doc/foo")) |> Result.isOk
```
[the remark this answers is not in the document]
Comment (esbenbjerre) [reply main/test/ca/uwaterloo/flix/library/Core/Io/TestFile.flix:36]: I have tried to, but I get `>> Undefined tag 'File'.` with the following implementation. I can't spot what the issue is.
```
use Core/Io/File.File;

instance Eq[File] {

    def eq(x: File, y: File): Bool = {
        import java.io.File.equals(##java.lang.Object);
        let File(xJ) = x;
        let File(yJ) = y;
        xJ.equals(yJ)
    }

    def neq(x: File, y: File): Bool = not (x == y)

}
```

[the remark this answers is not in the document]
Comment (esbenbjerre) [reply main/test/ca/uwaterloo/flix/library/Core/Io/TestFile.flix]: Nevermind I think the issue is that I was missing `use Core/Io/File.File.File;`. I'll see if I can wrap it up.

**Comment by magnus-madsen** [comment]
!!!benchmark

File: main/src/library/Set.flix [hunk 2]
```diff
@@ -37,6 +37,15 @@ instance Hash[Set[a]] with [a : Hash] {
         Set.foldLeft((acc, x) -> acc * 37 + Hash.hash(x), 7793, xs)
 }
 
+instance Eq[Set[a]] with[a : Eq] {
+
+    def eq(xs: Set[a], ys: Set[a]): Bool =
+        Set.forall(x -> Set.memberOf(x, ys), xs) and Set.forall(y -> Set.memberOf(y, xs), ys)
```
Comment (magnus-madsen) [inline main/src/library/Set.flix:43]: isSubSet?

**Comment by magnus-madsen** [comment]
I had one comment, but otherwise this looks ready for merge, yes?

File: main/src/library/Set.flix (hunk 2 above)
Comment (esbenbjerre) [reply main/src/library/Set.flix]: I see your point ;)

**Comment by esbenbjerre** [comment]
> I had one comment, but otherwise this looks ready for merge, yes?

Yes should be ready.

**Comment by magnus-madsen** [comment]
Thanks. I am *really* have we have this merged in.

**Pull request merged**